wallet/reservation.c

```diff
@@ -0,0 +1,282 @@
+		}
+	}
+
+	response = json_stream_success(cmd);
+	json_array_start(response, "utxos");
+	json_add_utxos(cmd->ld->wallet, response, utxos);
+	json_array_end(response);
+	return command_success(cmd, response);
+}
+
+static const struct json_command reserveinputs_command = {
+	"reserveinputs",
+	"bitcoin",
+	json_reserveinputs,
+	"Reserve utxos",
+	false
+};
+AUTODATA(json_command, &reserveinputs_command);
+
+static struct command_result *json_unreserveinputs(struct command *cmd,
+						   const char *buffer,
+						   const jsmntok_t *obj UNNEEDED,
+						   const jsmntok_t *params)
+{
+	struct json_stream *response;
+	struct utxo **utxos;
+
+	if (!param(cmd, buffer, params,
+		   p_req("utxos", param_utxos_any, &utxos),
+		   NULL))
+		return command_param_failed();
+
+	for (size_t i = 0; i < tal_count(utxos); i++) {
+		if (utxos[i]->status == output_state_reserved)
+			wallet_unreserve_utxo(cmd->ld->wallet, utxos[i]);
+	}
+
+	response = json_stream_success(cmd);
+	json_array_start(response, "utxos");
+	json_add_utxos(cmd->ld->wallet, response, utxos);
+	json_array_end(response);
+
+	return command_success(cmd, response);
+}
+
+static const struct json_command unreserveinputs_command = {
+	"unreserveinputs",
+	"bitcoin",
+	json_unreserveinputs,
+	"Unreserve utxos",
+	false
+};
+AUTODATA(json_command, &unreserveinputs_command);
+
+static size_t utxos_weight(struct utxo **utxos)
+{
+	size_t weight;
+
+	weight = bitcoin_tx_core_weight(tal_count(utxos), 0);
+
+	for (size_t i = 0; i < tal_count(utxos); i++)
+		weight += utxo_spend_weight(utxos[i]);
+
+	return weight;
+}
+
+/* Returns how much more we need, or 0 if we have enough */
+static bool satisfied(u32 feerate_per_kw,
+		      struct amount_sat required,
+		      struct utxo **utxos,
+		      struct amount_sat *diff)
+{
+	struct amount_sat inamount;
+
+	/* If they ask for a huge amount, we'll never be satisfied anyway */
+	if (!amount_sat_add(&required, required,
+			    amount_tx_fee(feerate_per_kw, utxos_weight(utxos))))
+		return false;
+
+	inamount = AMOUNT_SAT(0);
+	for (size_t i = 0; i < tal_count(utxos); i++) {
+		/* If we ever allow them to specify random (unvetted) utxos,
+		 * this could happen.  We're satisfied in this case, I guess. */
+		if (!amount_sat_add(&inamount, inamount, utxos[i]->amount))
+			return true;
+	}
+
+	/* If inamount >= required, this succeeds. */
+	if (amount_sat_sub(diff, inamount, required))
+		return true;
+
+	if (!amount_sat_sub(diff, required, inamount))
+		abort();
+	return false;
+}
+
+static struct command_result *json_fundinputs(struct command *cmd,
+					      const char *buffer,
+					      const jsmntok_t *obj UNNEEDED,
+					      const jsmntok_t *params)
+{
+	struct json_stream *response;
+	struct utxo **utxos;
+	u32 *feerate_per_kw;
+	u32 *minconf;
+	struct amount_sat *amount, diff;
+
+	if (!param(cmd, buffer, params,
+		   p_req("satoshis", param_sat_or_all, &amount),
+		   p_req("feerate", param_feerate, &feerate_per_kw),
```

Comment:
Should we not specify as well the number and type of expected outputs?  To get a better idea of how much fees will end up being?